src/persist-types/src/part.rs

```diff
@@ -0,0 +1,551 @@
+            ));
+        }
+        if self.len != self.diff.len() {
+            return Err(format!(
+                "diff col len {} didn't match part len {}",
+                self.diff.len(),
+                self.len()
+            ));
+        }
+        // TODO: Also validate the col types match schema.
+        Ok(())
+    }
+}
+
+/// An in-progress columnar constructor for one blob's worth of data.
+#[derive(Debug, Default)]
+pub struct PartBuilder {
+    key: Vec<(String, DynColumnMut)>,
+    val: Vec<(String, DynColumnMut)>,
+    ts: Vec<i64>,
+    diff: Vec<i64>,
+}
+
+impl PartBuilder {
+    /// Returns a new PartBuilder with the given schema.
+    pub fn new<K, KS: Schema<K>, V, VS: Schema<V>>(key_schema: &KS, val_schema: &VS) -> Self {
+        let key = key_schema
+            .columns()
+            .into_iter()
+            .map(|(name, data_type)| (name, DynColumnMut::new_untyped(&data_type)))
+            .collect();
+        let val = val_schema
+            .columns()
+            .into_iter()
+            .map(|(name, data_type)| (name, DynColumnMut::new_untyped(&data_type)))
+            .collect();
+        let ts = Vec::new();
+        let diff = Vec::new();
+        PartBuilder { key, val, ts, diff }
+    }
+
+    /// Returns a [ColumnsMut] for the key columns.
+    pub fn key_mut<'a>(&'a mut self) -> ColumnsMut<'a> {
+        ColumnsMut {
+            cols: self
+                .key
+                .iter_mut()
+                .map(|(name, col)| (name.as_str(), col))
+                .collect(),
+        }
+    }
+
+    /// Returns a [ColumnsMut] for the val columns.
+    pub fn val_mut<'a>(&'a mut self) -> ColumnsMut<'a> {
+        ColumnsMut {
+            cols: self
+                .val
+                .iter_mut()
+                .map(|(name, col)| (name.as_str(), col))
+                .collect(),
+        }
+    }
+
+    /// Adds a single timestamp and diff.
+    ///
+    /// TODO: Feels like there's some better way to model this.
+    pub fn push_ts_diff(&mut self, ts: i64, diff: i64) {
+        self.ts.push(ts);
+        self.diff.push(diff);
+    }
+
+    /// Completes construction of the [Part].
+    pub fn finish(self) -> Result<Part, String> {
+        let key = self
+            .key
+            .into_iter()
+            .map(|(name, col)| (name, col.finish_untyped()))
+            .collect();
+        let val = self
+            .val
+            .into_iter()
+            .map(|(name, col)| (name, col.finish_untyped()))
+            .collect();
+        let ts = Buffer::from(self.ts);
+        let diff = Buffer::from(self.diff);
+
+        let len = diff.len();
+        let part = Part {
+            len,
+            key,
+            val,
+            ts,
+            diff,
+        };
+        let () = part.validate()?;
+        Ok(part)
+    }
+}
+
+/// Hack to make things work with `Arc<dyn Any>::downcast_ref`.
+#[derive(Debug)]
+struct DynColumnRef(DataType, Arc<dyn Any>);
```

Comment:
ran into this pretty quickly while playing around with things -- we want this to be `dyn Any + Send` (and the associated changes on `Data::Mut` and `Data::Col` etc) so we can use a part builder inside a future